File: backend/app/routers/learning.py
"""
Learning State & Adaptive Scheduler endpoints.
Implements SM-2 spaced repetition for personalized review scheduling.
"""

from datetime import datetime, timezone, timedelta
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from sqlalchemy import func
from pydantic import BaseModel

from app.auth import get_current_user, require_student
from app.database import get_db
from app.models import User, TopicMastery, Topic, Subject, StudySession, TestAttempt
from app.services.scheduler import update_learning_state, quality_from_score, update_sm2

router = APIRouter()


class UpdateLearningRequest(BaseModel):
    topic_id: int
    score_percent: float  # 0-100


class SM2DebugRequest(BaseModel):
    repetitions: int = 0
    ease_factor: float = 2.5
    interval: int = 0
    quality: int = 3  # 0-5


@router.get("/due-today")
def get_due_today(
    current_user: User = Depends(require_student),
    db: Session = Depends(get_db),
):
    """Get topics due for review today based on SM-2 scheduling."""
    now = datetime.now(timezone.utc)

    # Get all mastery records for user
    records = db.query(TopicMastery, Topic).join(
        Topic, TopicMastery.topic_id == Topic.id
    ).filter(
        TopicMastery.user_id == current_user.id,
    ).all()

    due_topics = []
    for mastery, topic in records:
        # Calculate if due based on forgetting risk and last reviewed
        if not mastery.last_reviewed:
            due_topics.append({
                "topic_id": topic.id,
                "topic_name": topic.name,
                "subject": topic.subject.name if topic.subject else "",
                "mastery": mastery.mastery,
                "forgetting_risk": mastery.forgetting_risk,
                "priority": "high",
            })
            continue

        hours_since = (now - mastery.last_reviewed).total_seconds() / 3600
        interval_hours = 24 if mastery.forgetting_risk == "high" else (144 if mastery.forgetting_risk == "medium" else 336)

        if hours_since >= interval_hours:
            priority = "high" if mastery.forgetting_risk == "high" else ("medium" if mastery.forgetting_risk == "medium" else "low")
            due_topics.append({
                "topic_id": topic.id,
                "topic_name": topic.name,
                "subject": topic.subject.name if topic.subject else "",
                "mastery": mastery.mastery,
                "forgetting_risk": mastery.forgetting_risk,
                "hours_overdue": round(hours_since - interval_hours, 1),
                "priority": priority,
            })

    # Sort by priority
    priority_order = {"high": 0, "medium": 1, "low": 2}
    due_topics.sort(key=lambda x: priority_order.get(x.get("priority", "low"), 3))

    return {
        "due_count": len(due_topics),
        "topics": due_topics,
        "date": now.date().isoformat(),
    }


@router.post("/update")
def update_learning(
    payload: UpdateLearningRequest,
    current_user: User = Depends(require_student),
    db: Session = Depends(get_db),
):
    """Update learning state after a study session or quiz. Uses SM-2 algorithm."""
    # Validate topic exists
    topic = db.query(Topic).filter(Topic.id == payload.topic_id).first()
    if not topic:
        raise HTTPException(status_code=404, detail="Topic not found")

    result = update_learning_state(
        db=db,
        user_id=current_user.id,
        topic_id=payload.topic_id,
        score_percent=payload.score_percent,
    )

    return result


@router.get("/dashboard")
def learning_dashboard(
    current_user: User = Depends(require_student),
    db: Session = Depends(get_db),
):
    """User's learning progress overview — mastery per subject, streak, stats."""
    # Get all subjects for user's semester
    subjects = db.query(Subject).filter(Subject.semester == (current_user.semester or 6)).all()

    subject_progress = []
    for subj in subjects:
        topic_ids = db.query(Topic.id).filter(Topic.subject_id == subj.id).subquery()
        avg_mastery = db.query(func.avg(TopicMastery.mastery)).filter(
            TopicMastery.user_id == current_user.id,
            TopicMastery.topic_id.in_(topic_ids),
        ).scalar() or 0.0

        topics_total = db.query(Topic).filter(Topic.subject_id == subj.id).count()
        topics_started = db.query(TopicMastery).filter(
            TopicMastery.user_id == current_user.id,
            TopicMastery.topic_id.in_(topic_ids),
        ).count()

        subject_progress.append({
            "subject_id": subj.id,
            "subject_name": subj.name,
            "subject_code": subj.code,
            "avg_mastery": round(avg_mastery, 1),
            "topics_total": topics_total,
            "topics_started": topics_started,
        })

    # Study streak
    sessions = db.query(StudySession).filter(
        StudySession.user_id == current_user.id,
        StudySession.ended_at.isnot(None),
    ).order_by(StudySession.started_at.desc()).all()

    streak = 0
    if sessions:
        dates = sorted(set(s.started_at.date() for s in sessions), reverse=True)
        today = datetime.now(timezone.utc).date()
        if dates and (dates[0] == today or dates[0] == today - timedelta(days=1)):
            streak = 1
            for i in range(1, len(dates)):
                if (dates[i - 1] - dates[i]).days == 1:
                    streak += 1
                else:
                    break

    # Total stats
    total_study_minutes = db.query(func.sum(StudySession.duration_minutes)).filter(
        StudySession.user_id == current_user.id
    ).scalar() or 0

    tests_taken = db.query(TestAttempt).filter(
        TestAttempt.user_id == current_user.id,
        TestAttempt.is_completed == True,
    ).count()

    return {
        "user": current_user.full_name,
        "semester": current_user.semester,
        "streak_days": streak,
        "total_study_hours": round(total_study_minutes / 60, 1),
        "tests_taken": tests_taken,
        "subjects": subject_progress,
    }


@router.post("/sm2-calculate")
def sm2_calculate(payload: SM2DebugRequest):
    """Debug/test endpoint: calculate SM-2 output for given inputs."""
    new_reps, new_ef, new_interval = update_sm2(
        payload.repetitions, payload.ease_factor, payload.interval, payload.quality
    )
    return {
        "input": payload.model_dump(),
        "output": {
            "repetitions": new_reps,
            "ease_factor": round(new_ef, 2),
            "interval_days": new_interval,
            "next_review": (datetime.now(timezone.utc) + timedelta(days=new_interval)).isoformat(),
        },
    }

File: backend-ts/src/routes/learning.ts
import { Router, Response } from 'express';
import { prisma } from '../prisma';
import { authenticate, AuthRequest } from '../middleware/auth';
import { getCached, setCache } from '../cache';

const router = Router();

// GET /api/learning/due-today
router.get('/due-today', authenticate, async (req: AuthRequest, res: Response) => {
  try {
    const now = new Date();
    const cards = await prisma.spacedRepetitionCard.findMany({
      where: { userId: req.user!.id, nextReview: { lte: now } },
      include: { topic: { include: { subject: true } } },
      orderBy: { nextReview: 'asc' },
    });
    return res.json({
      due_count: cards.length,
      topics: cards.map(c => ({
        card_id: c.id, topic: c.topic.name, subject: c.topic.subject.name,
        ease_factor: c.easeFactor, interval: c.interval, next_review: c.nextReview,
      })),
      date: now.toISOString().split('T')[0],
    });
  } catch (err: any) {
    return res.status(500).json({ detail: err.message });
  }
});

// POST /api/learning/update
router.post('/update', authenticate, async (req: AuthRequest, res: Response) => {
  try {
    const { topic_id, score_percent } = req.body;
    if (!topic_id || score_percent === undefined) {
      return res.status(400).json({ detail: 'topic_id and score_percent required' });
    }

    const userId = req.user!.id;

    // Get or create mastery
    let mastery = await prisma.topicMastery.findUnique({ where: { userId_topicId: { userId, topicId: topic_id } } });
    if (!mastery) {
      mastery = await prisma.topicMastery.create({ data: { userId, topicId: topic_id, mastery: score_percent } });
    }

    // SM-2 update
    let card = await prisma.spacedRepetitionCard.findUnique({ where: { userId_topicId: { userId, topicId: topic_id } } });
    const quality = score_percent >= 95 ? 5 : score_percent >= 80 ? 4 : score_percent >= 60 ? 3 : score_percent >= 40 ? 2 : score_percent >= 20 ? 1 : 0;

    let newEF = 2.5, newInterval = 1, newReps = 0;
    if (card) {
      if (quality >= 3) {
        newReps = card.repetitions + 1;
        newInterval = newReps === 1 ? 1 : newReps === 2 ? 6 : Math.round(card.interval * card.easeFactor);
      }
      newEF = card.easeFactor + (0.1 - (5 - quality) * (0.08 + (5 - quality) * 0.02));
      newEF = Math.max(1.3, newEF);
    } else {
      newReps = quality >= 3 ? 1 : 0;
      newInterval = quality >= 3 ? 1 : 1;
    }

    const nextReview = new Date();
    nextReview.setDate(nextReview.getDate() + newInterval);

    if (card) {
      await prisma.spacedRepetitionCard.update({
        where: { id: card.id },
        data: { easeFactor: newEF, interval: newInterval, repetitions: newReps, nextReview, lastReview: new Date() },
      });
    } else {
      await prisma.spacedRepetitionCard.create({
        data: { userId, topicId: topic_id, easeFactor: newEF, interval: newInterval, repetitions: newReps, nextReview, lastReview: new Date() },
      });
    }

    // Update mastery
    const newMastery = Math.min(100, mastery.mastery * 0.7 + score_percent * 0.3);
    const risk = newInterval >= 14 ? 'low' : newInterval >= 6 ? 'medium' : 'high';
    await prisma.topicMastery.update({
      where: { id: mastery.id },
      data: { mastery: Math.round(newMastery * 10) / 10, forgettingRisk: risk, lastReviewed: new Date() },
    });

    // Log event
    await prisma.learningEvent.create({
      data: { userId, eventType: 'mastery_update', topicId: topic_id, payload: JSON.stringify({ score: score_percent, quality }) },
    });

    // Invalidate caches
    setCache(`dashboard:${userId}`, null, 0);
    setCache(`progress:${userId}`, null, 0);
    setCache(`activity:${userId}`, null, 0);

    return res.json({ topic_id, mastery: Math.round(newMastery * 10) / 10, quality, new_interval_days: newInterval, new_ease_factor: Math.round(newEF * 100) / 100, next_review: nextReview.toISOString() });
  } catch (err: any) {
    return res.status(500).json({ detail: err.message });
  }
});

// GET /api/learning/dashboard
router.get('/dashboard', authenticate, async (req: AuthRequest, res: Response) => {
  try {
    const userId = req.user!.id;
    const user = await prisma.user.findUnique({ where: { id: userId } });
    const mastery = await prisma.topicMastery.findMany({ where: { userId }, include: { topic: { include: { subject: true } } } });
    const sessions = await prisma.studySession.findMany({ where: { userId } });
    const attempts = await prisma.testAttempt.findMany({ where: { userId, isCompleted: true } });

    // Group by subject
    const subjectMap = new Map<string, { name: string; topics: any[]; avgMastery: number }>();
    for (const m of mastery) {
      const subName = m.topic.subject.name;
      if (!subjectMap.has(subName)) subjectMap.set(subName, { name: subName, topics: [], avgMastery: 0 });
      subjectMap.get(subName)!.topics.push({ name: m.topic.name, mastery: m.mastery, risk: m.forgettingRisk });
    }
    for (const [, sub] of subjectMap) {
      sub.avgMastery = sub.topics.reduce((s, t) => s + t.mastery, 0) / (sub.topics.length || 1);
    }

    // Calculate proper streak
    const now = new Date();
    let streak = 0;
    for (let i = 0; i < 365; i++) {
      const date = new Date(now);
      date.setDate(date.getDate() - i);
      const dayStr = date.toISOString().split('T')[0];
      const hasSession = sessions.some(s => s.startedAt.toISOString().startsWith(dayStr));
      if (hasSession) streak++;
      else if (i > 0) break;
    }

    return res.json({
      user: { name: user?.fullName, semester: user?.semester },
      semester: user?.semester || 6,
      streak_days: streak,
      total_study_hours: Math.round(sessions.reduce((s, ss) => s + ss.durationMinutes, 0) / 60 * 10) / 10,
      tests_taken: attempts.length,
      subjects: Array.from(subjectMap.values()),
    });
  } catch (err: any) {
    return res.status(500).json({ detail: err.message });
  }
});

// POST /api/learning/sm2-calculate
router.post('/sm2-calculate', authenticate, async (req: AuthRequest, res: Response) => {
  try {
    const { repetitions, ease_factor, interval, quality } = req.body;
    const q = Math.max(0, Math.min(5, quality || 3));
    let newReps: number, newInterval: number;

    if (q >= 3) {
      newReps = (repetitions || 0) + 1;
      newInterval = newReps === 1 ? 1 : newReps === 2 ? 6 : Math.round((interval || 1) * (ease_factor || 2.5));
    } else {
      newReps = 0;
      newInterval = 1;
    }

    let newEF = (ease_factor || 2.5) + (0.1 - (5 - q) * (0.08 + (5 - q) * 0.02));
    newEF = Math.max(1.3, newEF);

    return res.json({ repetitions: newReps, ease_factor: Math.round(newEF * 100) / 100, interval: newInterval, quality: q });
  } catch (err: any) {
    return res.status(500).json({ detail: err.message });
  }
});

export default router;
